cmd/clusterctl/internal/migrate/parser.go

```diff
@@ -0,0 +1,216 @@
+
+	var documents []Document
+	index := 0
+
+	for {
+		raw, err := yamlReader.Read()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to read YAML document")
+		}
+
+		trimmed := bytes.TrimSpace(raw)
+		if len(trimmed) == 0 {
+			continue
+		}
+
+		doc, err := p.parseDocument(trimmed, raw, index)
+		if err != nil {
+			// Treat parsing failures as unsupported documents but keep raw bytes.
+			documents = append(documents, Document{
+				Object: nil,
+				Raw:    p.ensureTrailingNewline(raw),
+				GVK:    schema.GroupVersionKind{},
+				Type:   ResourceTypeUnsupported,
+				Index:  index,
+			})
+		} else {
+			documents = append(documents, doc)
+		}
+		index++
+	}
+
+	return documents, nil
+}
+
+// SerializeYAMLStream writes documents back out as a multi-document YAML stream.
+func (p *YAMLParser) SerializeYAMLStream(docs []Document, writer io.Writer) error {
+	if len(docs) == 0 {
+		return nil
+	}
+
+	for i, doc := range docs {
+		// Add document separator before each document except the first
+		if i > 0 {
+			if _, err := io.WriteString(writer, "---\n"); err != nil {
+				return errors.Wrap(err, "failed to write document separator")
+			}
+		}
+
+		if doc.Object != nil {
+			buf := &bytes.Buffer{}
+			if err := p.yamlSerializer.Encode(doc.Object, buf); err != nil {
+				return errors.Wrapf(err, "failed to encode document at index %d", doc.Index)
+			}
+			if _, err := writer.Write(p.ensureTrailingNewline(buf.Bytes())); err != nil {
+				return errors.Wrapf(err, "failed to write document at index %d", doc.Index)
+			}
+			continue
+		}
+
+		if _, err := writer.Write(p.ensureTrailingNewline(doc.Raw)); err != nil {
+			return errors.Wrapf(err, "failed to write raw document at index %d", doc.Index)
+		}
+	}
+
+	return nil
+}
+
+func (p *YAMLParser) parseDocument(trimmed []byte, original []byte, index int) (Document, error) {
+	obj := &unstructured.Unstructured{}
+	_, gvk, err := p.unstructuredDecoder.Decode(trimmed, nil, obj)
+	if err != nil || gvk == nil || gvk.Empty() {
+		return Document{}, errors.New("unable to decode document")
+	}
+
+	resourceType := p.classifyResource(*gvk)
+
+	var runtimeObj runtime.Object
+	if resourceType == ResourceTypeCoreV1Beta1 || resourceType == ResourceTypeOtherCAPI {
+		if typedObj, _, err := p.typedDecoder.Decode(trimmed, gvk, nil); err == nil {
+			runtimeObj = typedObj
+		} else {
+			runtimeObj = obj
+		}
+	} else {
+		runtimeObj = obj
+	}
+
+	return Document{
+		Object: runtimeObj,
+		Raw:    p.ensureTrailingNewline(original),
+		GVK:    *gvk,
+		Type:   resourceType,
+		Index:  index,
+	}, nil
+}
+
+func (p *YAMLParser) ensureTrailingNewline(content []byte) []byte {
+	if len(content) == 0 {
+		return content
+	}
+	if content[len(content)-1] != '\n' {
+		content = append(content, '\n')
+	}
+	return content
+}
+
+func (p *YAMLParser) classifyResource(gvk schema.GroupVersionKind) ResourceType {
+	if p.isCoreCAPIGroup(gvk.Group) {
+		if gvk.Version == "v1beta1" {
```

Comment:
best to check from the input GVK i mention above instead of hardcoding known versions in string literals.